Automatically size the Pixel FED buffer used by SiPixelRawToClusterCUDA

fwyzard opened the pull request:
#### PR description:

Automatically resize the Pixel FED buffer used by `SiPixelRawToClusterCUDA` to construct a contiguous copy of the FED data to be copied to the GPU.

The value of the `MaxFEDWords` parameter is not used. It is kept for backwards compatibility with older configurations, but it will not be written to the cfi files and will be ignored if present.

#### PR validation:

Tested on events from run 355872 that would crash with a segmentation fault or raise an exception, because the Pixel had a larger occupancy than expected.
With these changes all crashes disappear.

#### If this PR will be backported please specify to which release cycle the backport is meant for:

To be backported to 12.4.x for data taking.

fwyzard committed:
Automatically resize the Pixel FED buffer

fwyzard commented:
@makortel @VinInn @AdrianoDee what do you think ?

fwyzard commented:
please test

fwyzard commented:
enable gpu

fwyzard commented:
please test

fwyzard commented:
type bugfix

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38868/31284

ERROR: Build errors found during clang-tidy run.
```
RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc:256:10: error: no member named 'LogWrning' in namespace 'edm' [clang-diagnostic-error]
    edm::LogWrning("SiPixelRawToClusterCUDA")
         ^
Suppressed 1463 warnings (1462 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

fwyzard commented:
please test

fwyzard committed:
Mark variables used only inside an assert as [[maybe_unused]]

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38868/31286

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu  modified in PR(s): #37359, #38761
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h  modified in PR(s): #37359, #38761
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h  modified in PR(s): #37359, #38761, #35713

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- RecoLocalTracker/SiPixelClusterizer (**reconstruction**)


@jpata, @clacaputo can you please review it and eventually sign? Thanks.
@mtosi, @VourMa, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @OzAmram, @ferencek, @dkotlins, @gpetruc, @mmusich, @threus, @tvami this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c33ceb/26474/summary.html
**COMMIT**: 79c5e09c991f419a3c025281588383f4446bfc61
**CMSSW**: CMSSW_12_5_X_2022-07-26-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38868/26474/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - Reco comparison had  3 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 36
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19840
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_5_X_2022-07-26-2300+c33ceb/51817/triggerResults): found differences in 1 / 3 workflows


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3667670
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3667646
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 210 log files, 47 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-26-2300+c33ceb/51819/triggerResults): no differences found

makortel commented:
> what do you think ?

Looks good to me. I only wonder if anyone would notice the warning and act based on that? (because of my ignorance on how warnings are actually tracked in various applications)

mmusich commented on RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h:
```diff
@@ -30,7 +30,8 @@ namespace gpuClustering {
     constexpr int startBPIX2 = isPhase2 ? phase2PixelTopology::layerStart[1] : phase1PixelTopology::layerStart[1];
-    constexpr int nMaxModules = isPhase2 ? phase2PixelTopology::numberOfModules : phase1PixelTopology::numberOfModules;
+    [[maybe_unused]] constexpr int nMaxModules =
+        isPhase2 ? phase2PixelTopology::numberOfModules : phase1PixelTopology::numberOfModules;
```
I suppose these are to silence the compilation warnings.
Out of curiosity, using them I an assertion one line below doesn't count as "use"?

fwyzard replied:
Correct.

No, because on GPUs we disable assertions, so the content of `assert(...)` is never seen by the compiler.

mmusich replied:
Thanks

fwyzard commented:
> I only wonder if anyone would notice the warning and act based on that? (because of my ignorance on how warnings are actually tracked in various applications)

In my experience LogWarnings are usually ignored, until somebody is bothered enough by them and request silencing them :-)

fwyzard commented:
In fact, once I have some time (_i.e._ am not on night shifts) I would try to benchmark
  - this PR
  - replacing the write-combined memory with standard pinned memory
  - allocating the buffer on every event using the host caching allocator

which would remove the need for a preallocated buffer entirely.

clacaputo commented:
> * this PR
> * replacing the write-combined memory with standard pinned memory
> * allocating the buffer on every event using the host caching allocator

Hi @fwyzard , are these action items expected to converge in this PR?

fwyzard commented:
@clacaputo I'm not sure on the time scale, because I'm investigating a couple other GPU/HLT-related problems right now.

Let me check with the other HLT people so I can set priorities and work on what is most urgent.

fwyzard commented:
hold

OK, looks like this PR is not _terribly_ urgent, so I will try to run the various benchmarks before merging it.

cmsbuild commented:
Pull request has been put on hold by  @fwyzard
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

VinInn commented:
I think that current standard is 
- allocating the buffer on every event using the host caching allocator

fwyzard commented:
Here are some benchmarks, running only the pixel tracks reconstruction (including copies from gpu to cpu and conversion to legacy format) over 10k events from run 355990 on a full HLT production node (2x64 cores, 2x T4 gpus) using 8 jobs with 32 thread each.

description | throughput
---|---
original code | 1713.3 ±  14.6 ev/s
minimal fix (this PR as of https://github.com/cms-sw/cmssw/pull/38868/commits/d1909af72c03c91225630981cb512ba51cb20b70) | 1713.2 ±  15.7 ev/s
using non-write-combine memory | 1715.3 ±  15.1 ev/s
allocating the `WordFedAppender` for each event | 1721.2 ±  15.7 ev/s

Since reallocating the `WordFedAppender` does not cause any slowdown, and simplifies the code a bit, I'll push this version.

fwyzard committed:
Autodetect the Pixel FED size

Allocate a WordFedAppender for each event using non-write-conbined,
pinned memory (generally from the host caching allocator) based on the
actual size of the FEDs.

This makes the MaxFEDWords configuration parametere obsolete. It is
still kept for backwards compatibility with older configurations, but it
will not be written to the cfi files and will be ignored if present.

fwyzard commented:
please test

fwyzard commented:
@clacaputo this should be OK to go

fwyzard commented:
unhold

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38868/31335

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu  modified in PR(s): #37359, #38761
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h  modified in PR(s): #37359, #38761
  - File  RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h  modified in PR(s): #37359, #38761, #35713

cmsbuild commented:
Pull request #38868 was updated. @jpata, @clacaputo can you please check and sign again.

fwyzard commented:
urgent

clacaputo commented:
Is the test stuck? It seems it is still running after 3 days. I'll try to retrigger it

clacaputo commented:
@cmsbuild please test

clacaputo commented:
enable gpu

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c33ceb/26559/summary.html
**COMMIT**: c8fae54b8cb123d91a8639cd8851e949c75661a5
**CMSSW**: CMSSW_12_5_X_2022-08-01-1100/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38868/26559/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 1094
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 18782
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_5_X_2022-08-01-1100+c33ceb/51893/triggerResults): found differences in 2 / 3 workflows


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3669004
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3668980
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 210 log files, 47 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-01-1100+c33ceb/51895/triggerResults): no differences found

clacaputo commented:
Hi @fwyzard , I can see some minor [reco differences](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_5_X_2022-08-01-1100+c33ceb/51893/validateJR/all_OldVSNew_TTbar14TeV2021pixel3GPUwf11634p506/) (maybe fluctuations) for GPU workflow `11634.506`. Can they be classified as some GPU-effect fluctuations? I can't see the same effect in the 124X backport #38902

fwyzard commented:
Hi @clacaputo , this PR should be purely technical, with no visible effects on the reconstructed quantities (apart from fixing some crashes...).
I think the discrepancies seem small enough that we can ascribe them to the usual random fluctuations in GPU code.

clacaputo commented:
+reconstruction

- technical PR
- reco changes observed are small enough to be considered fluctuations in GPU code

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
